### PR TITLE
8.0 website parametric snippets

### PR DESCRIPTION
**Pull request by gfcapalbo**
Adds Parametric tag to Qweb views,  allowing for parametric dynamic widgets.

### CODE REVIEW

**Comment by hbrunn** [comment]
your module name violates https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#modules
Given you have to change the name anyways, how about `parameterized`?

File: website_parametric_snippets/README.rst [hunk 1]
```diff
@@ -0,0 +1,101 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+    :alt: License: AGPL-3
+
+===========================
+website_parametric_snippets
```
Comment (hbrunn) [inline website_parametric_snippets/README.rst]: use human readable name

File: website_parametric_snippets/README.rst [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+===========================
+
+This module adds extensions to view_view in order to render available
```
Comment (hbrunn) [inline website_parametric_snippets/README.rst]: what does 'view_view' mean?

File: website_parametric_snippets/README.rst [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+===========================
+website_parametric_snippets
+===========================
+
+This module adds extensions to view_view in order to render available 
+to the programmer a new tag in qweb views: "t-ignore-branding".
+Everything inside this tag will be ignored by the qweb evaluator, 
+allowing the code to be evaluated on the fly.
+
+The final result of this will be dynamic parameters in your widgets.
```
Comment (hbrunn) [inline website_parametric_snippets/README.rst]: explain here why we would want that

File: website_parametric_snippets/README.rst [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+==================================
+
+File minimal.xml and minimal.js in subdirectory docs contain 
+a basic example of usage for this module.
+
+
+Examples of modules that use this tag (will be updated)
+=======================================================
+website_snippet_blog_display_post
+website_twitter_no_ext_links
```
Comment (hbrunn) [inline website_parametric_snippets/README.rst]: those want to be links

File: website_parametric_snippets/README.rst [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+Images
+------
+
+* Odoo Community Association: `Icon <https://github.com/OCA/maintainer-tools/blob/master/template/module/static/description/icon.svg>`_.
+
+Contributors
+------------
+
+* gfcapalbo <giovanni@therp.nl>
```
Comment (hbrunn) [inline website_parametric_snippets/README.rst]: use full names

File: website_parametric_snippets/README.rst [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+Bugs are tracked on `GitHub Issues <https://github.com/OCA/website_panam_oca/issues>`_.
+In case of trouble, please check there if your issue has already been reported.
+If you spotted it first, help us smashing it by providing a detailed and welcomed feedback
+`here <https://github.com/OCA/website_panam_oca/issues/new?body=module:%20website_parametric_snippets%0Aversion:%208.0%0A%0A**Steps%20to%20reproduce**%0A-%20...%0A%0A**Current%20behavior**%0A%0A**Expected%20behavior**>`_.
+
+Credits
+=======
+
+Thanks to Hbrunn <hbrunn@therp.nl> for the parametric t tag idea
```
Comment (hbrunn) [inline website_parametric_snippets/README.rst]: isn't the contributor part enough?

File: website_parametric_snippets/__openerp__.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
+# © 2016 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+{
+    "name": "website_parametric_snippets",
```
Comment (hbrunn) [inline website_parametric_snippets/__openerp__.py]: use a human readable name

File: website_parametric_snippets/__openerp__.py [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+    "version": "8.0.1.0.0",
+    "author": "Therp BV,Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "category": "",
```
Comment (hbrunn) [inline website_parametric_snippets/__openerp__.py]: choose a category, probably `Hidden/Dependency`

File: website_parametric_snippets/__openerp__.py [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    "summary": "",
+    "depends": [
+    ],
```
Comment (hbrunn) [inline website_parametric_snippets/__openerp__.py]: website or base

File: website_parametric_snippets/__openerp__.py [hunk 10]
```diff
@@ -0,0 +1,32 @@
+# -*- coding: utf-8 -*-
+# © 2016 Therp BV <http://therp.nl>
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+{
+    "name": "website_parametric_snippets",
+    "version": "8.0.1.0.0",
+    "author": "Therp BV,Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "category": "",
+    "summary": "",
```
Comment (hbrunn) [inline website_parametric_snippets/__openerp__.py]: add a summary

File: website_parametric_snippets/__openerp__.py [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+    ],
+    "pre_init_hook": False,
+    "post_init_hook": False,
+    "uninstall_hook": False,
+    "auto_install": False,
+    "installable": True,
+    "application": False,
+    "external_dependencies": {
+        'python': [],
+    },
```
Comment (hbrunn) [inline website_parametric_snippets/__openerp__.py]: delete empty keys

File: website_parametric_snippets/models/view_view.py [hunk 12]
```diff
@@ -0,0 +1,14 @@
+# -*- coding: utf-8 -*-
```
Comment (hbrunn) [inline website_parametric_snippets/models/view_view.py]: rename file to class name

**Comment by lmignon** [comment]
The same idea from @hbrunn has already been proposed in https://github.com/OCA/website/pull/141/files
To be complete you also need to extend `website.qweb.field.html` to render `t-call` elements inserted by snippets using the parametic approach and stored on the related model itself. (for exemplet blog.post)...
see https://github.com/OCA/website/pull/141/files#diff-3d8e1a5ab997cb1092e70c7fa7fdc25fR9

File: website_parametric_snippets/docs/minimal.xml [hunk 13]
```diff
@@ -0,0 +1,55 @@
+<openerp>
+    <data>
```
Comment (hbrunn) [inline website_parametric_snippets/docs/minimal.xml]: Include those files in your manifest in the demo key, then people can play with it on runbot

**Comment by gfcapalbo** [comment]
thank you @lmignon  checking out the PR

**Comment by yajo** [comment]
I'm getting 0 changed files here...

File: website_parameterized_snippet/README.rst [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+===========================
+
+This module adds extensions in order to render available 
+to the programmer a new tag in qweb views: "t-ignore-branding".
+Everything inside this tag will be ignored by the qweb evaluator, 
+allowing the code to be evaluated on the fly.
+
+The final result of this will be dynamic parameters in your widgets.
+This would allow widgets to be defined as "Show the last 5 entries of
+this model"  or "Show all the data that has an attribute corrisponding
```
Comment (pedrobaeza) [inline website_parameterized_snippet/README.rst]: s/corrisponding/corresponding

**Comment by gfcapalbo** [comment]
@lmignon  added your improvements.

**Comment by gfcapalbo** [comment]
Closing this PR. Will propose a MR to @lmignon 's module and make all my snippets depend on it.

**Comment by gfcapalbo** [comment]
@lmignon  Wouldn't it be nice to split   website_node_branded  in it's own PR?  I have some   documentation to contribute i think it would be cleaner./

**Comment by lmignon** [comment]
@gfcapalbo I propose that you keep your PR since the idea comes from your company. https://github.com/OCA/website/pull/141 could be rebased on your PR to allow 'Therp BV' to keep the copyright on this nice idea.

**Comment by gfcapalbo** [comment]
@lmignon  agreed. will make this branch pass tests and leave PR open.

**Comment by lmignon** [comment]
@gfcapalbo @hbrunn Thank you for this great idea that open a wide range of possibilities to create configurable snippets providing dynamic contents . 
Can you take a look to the errors in travis?
https://travis-ci.org/OCA/website/jobs/147444147#L469
https://travis-ci.org/OCA/website/jobs/147444148#L1236

**Comment by hbrunn** [comment]
THERE ARE NO COPYRIGHTS ON IDEAS
(or at least shouldn't be. Sorry for not writing anything more sensible, but I find this statement important)

**Comment by hbrunn** [comment]
@gfcapalbo the runbot error probably is caused by the renaming and you have to adapt file names too

**Comment by lmignon** [comment]
For the record here it's the PR where I discovered the concept of parametric template in action.. https://github.com/OCA/website/pull/125/files ...

**Comment by hbrunn** [comment]
@gfcapalbo you still need to fix filenames here: https://travis-ci.org/OCA/website/jobs/147444148#L1236 and by now, there's also a linter for js: https://travis-ci.org/OCA/website/jobs/147444147#L475

**Comment by gfcapalbo** [comment]
@lmignon  @hbrunn trouble with the files in the "minimal" subdirectory.
I put those in to promote the usage of this module.... as minimal examples.\
I will remove them, they cause lint JS errors and will just keep on adding the modules that use this to the list.
Please do the same when you write a module that uses the  t-ignore-branding tag.

**Comment by hbrunn** [comment]
@gfcapalbo I think it's a good thing to fix up the minimal files and add a file pulling them as demo data. then people can test this on runbot

**Comment by gfcapalbo** [comment]
@hbrunn @lmignon  branch passing, can remove "needs fixing", I hope this will pass soon, it is holding up:

https://github.com/OCA/website/pull/238
https://github.com/OCA/website/pull/227

@lmignon  will you base https://github.com/OCA/website/pull/141/files on this too? It would make your PR also smaller.  If you intend to do so , please add in the README of this module your module name to the list of modules that use this.

**Comment by lmignon** [comment]
@gfcapalbo I already have asked to my colleague to rebase his job on your PR https://github.com/OCA/website/pull/141#issuecomment-235256335

**Comment by lmignon** [comment]
:+1: (Code review + tests done with the same code but in an other module)

**Comment by hbrunn** [comment]
still a pity with the demo files imho, but not blocking: :+1:

**Comment by gfcapalbo** [comment]
@hbrunn thanks

**Comment by lasley** [comment]
👍

**Pull request merged**